### PR TITLE
[ONNX] Bench torch.onnx.dynamo_export and torch.onnx.export under dynamo bench

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103983
* __->__ #103135

- Extend dynamo bench interface with '--compilers onnx' and '--compilers dynamo-onnx'
- ONNX bench exports model to onnx and runs in ONNX Runtime.
- Introduce error aggregation and report.
- Scripts to build ONNX deps and running ONNX bench.
- Huggingface accuracy check workaround for ONNX.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -2183,6 +2615,17 @@ def get_example_inputs(self):
     group.add_argument(
         "--xla", action="store_true", help="Compare TorchXLA to eager PyTorch"
     )
+    group.add_argument(
+        "--onnx",
```
Comment (thiagocrepaldi) [inline benchmarks/dynamo/common.py:2619]:
```suggestion
        "--torchscript-onnx",
```
Similar to the `OnnxModel` and `OnnxModelV2` comment, a more intuitive description would be nice. Although `--dynamo-onnx` is clear `--onnx` allows a lot of imagination, especially for those who are newer to pytorch and never met the torch script friend

**Review by thiagocrepaldi — COMMENTED** [review]
Overall, it looks very good.

The patching bit feels not clear why the extra mangling is needed. 
Several NITs mainly because this will be plugged outside the `torch.onnx` namespace and more clarification might make things easier

File: benchmarks/dynamo/_onnx/bench.sh [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    echo "  --device <cpu|cuda>   Specify the device to use: cpu or cuda"
+    echo "  --quick               Optional flag to run on small subset of ~3 models. Helps in debugging."
+    echo "  --filter <model_name> Optional flag to filter benchmarks with regexp on model names. e.g. --filter resnet50"
+    echo "  -h                    Display this help message"
+}
+
+# Parse command-line arguments
+quick=""
+filter=""
+while [[ $# -gt 0 ]]
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/bench.sh:22]: If undefined it errors out in the later check for required args.

File: benchmarks/dynamo/_onnx/bench.sh [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+if [[ $device != "cpu" && $device != "cuda" ]]; then
+    echo "Invalid device specified: $device"
+    show_usage
+    exit 1
+fi
+
+pushd "../../../"
+
+log_folder=".logs/onnx_bench"
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/bench.sh:68]: good catch, the rm -rf && mkdir is part of the bench script this script is calling, so we are safe.

File: benchmarks/dynamo/_onnx/bench.sh [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+# NOTE: use 'nohup' and add '&' to the end to prevent script stopping due to terminal timeout.
+
+set -e
+
+function show_usage {
+    echo "Usage: bench.sh --device <cpu|cuda> [--quick] [--filter <model_name>] [-h]"
+    echo "Options:"
+    echo "  --device <cpu|cuda>   Specify the device to use: cpu or cuda"
+    echo "  --quick               Optional flag to run on small subset of ~3 models. Helps in debugging."
+    echo "  --filter <model_name> Optional flag to filter benchmarks with regexp on model names. e.g. --filter resnet50"
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/bench.sh:15]: This is a good point. The model names are defined in their respective benchmark suites. Will emphasize on this when we create/reference doc for adding models to torchbench.

In other cases, benchmark report shows the list of model names.

File: benchmarks/dynamo/_onnx/patch.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+from torch.utils import _pytree as pytree
+
+
+def patch_non_tensor_outputs(correct_result, new_result, fp64_outputs):
+    try:
+        from transformers import modeling_outputs
+    except ImportError:
+        has_transformers = False
+    else:
+        has_transformers = True
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/patch.py:10]: Not following the question on discrepancy.

Are you worried that benchmark runs with huggingface suite, yet this import fails? We will get accuracy check failed for these models if that happens.

File: benchmarks/dynamo/_onnx/reporter.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+if __name__ == "__main__":
+    args = parse_args()
+
+    output_dir = pathlib.Path(args.output_dir)
+    compilers = args.compilers
+    assert compilers, "Must specify at least one compiler"
+    suites = args.suites
+    assert suites, "Must specify at least one suite"
+    device = args.device
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/reporter.py:383]: parser would reject if otherwise

File: benchmarks/dynamo/_onnx/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
+This folder contains python scripts to extend the dynamo benchmarking framework for benchmarking ONNX export. Bash scripts to run the benchmark and generate markdown reports are also provided.
+
+# Usage
+
+## Setup
+
+It is recommended to create a fresh python environment, clone and build PyTorch from source.
+
+It is expected that PyTorch is already built. Run the following script to install the benchmark dependencies:
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/README.md:9]: Guess I should ditch this line `It is expected that PyTorch is already built.`

[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/README.md:9]: But it is not contradicting, the first line asks to build pytorch from source. :)

File: benchmarks/dynamo/common.py [hunk 8]
```diff
@@ -975,6 +1098,300 @@ def wrapper(self, *args, **kwargs) -> Any:
...
+        pt_outputs = [
+            torch.from_numpy(ort_output).to(current_device)
+            for ort_output in ort_outputs
+        ]
+        if len(pt_outputs) == 1:
+            return pt_outputs[0]
+        return pt_outputs
+
+
+class OnnxModelV2(OnnxModel):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/common.py:1289]: No strong opinion, the origin is explained in the comment more precisely. I'll make it docstring and rename the classes.

File: benchmarks/dynamo/_onnx/patch.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
+from torch.utils import _pytree as pytree
+
+
+def patch_non_tensor_outputs(correct_result, new_result, fp64_outputs):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/patch.py:4]: adding docstring

File: benchmarks/dynamo/_onnx/reporter.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+class ErrorAggregator(object):
+    """
+    Collect and group error messages for report at the end
+
+    Copied and modified from pytorch-jit-paritybench/paritybench/reporting.py
+    """
+
+    # If these keywords are found in an error message, it won't be aggregated with other errors,
+    # unless it is an exact match.
+    _EXACT_MATCH_KEYWORDS = [
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/reporter.py:34]: done.

File: benchmarks/dynamo/_onnx/patch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply benchmarks/dynamo/_onnx/patch.py:16]: Yes, I was worried that the accuracy would change depending on `has_transformers` value.
Which brings the question whether we should make `transformers` mandatory to have uniform results.

Imagine that for 6 months `transformers` is not installed and we log one result. Next, we install it on the environment and the accuracy jumps (for bad and good). If that happens (my original question was if that is a possibility), the accuracy discrepancy would be because of environment change, not any improvement or regression on the exporter.

File: benchmarks/dynamo/_onnx/bench.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply benchmarks/dynamo/_onnx/bench.sh:22]: Np, that is just a NIT. 
"Referencing before assigning" is usually a compile/runtime error in many languages, so we end up always declaring it.

File: benchmarks/dynamo/_onnx/bench.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply benchmarks/dynamo/_onnx/bench.sh:15]: Awesome, we don't need that to get started benching our stuff

**Review by thiagocrepaldi — APPROVED** [review]
LGTM

How about adding `benchmarks/dynamo/_onnx/` to `https://github.com/pytorch/pytorch/blob/main/CODEOWNERS#L59` so that we can make updates to the ONNX part of the bench code in the future?

File: benchmarks/dynamo/_onnx/patch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply benchmarks/dynamo/_onnx/patch.py:16]: I see, in that case that's fine, because if `has_transformers` is False, it means no transformers models were tested, so that part of patch is irrelevant.

**Comment by msaroufim** [comment]
TBH I'm not that crazy about having the `benchmarks/_dynamo/onnx` folder checked into core, it feels more appropriate to have this be part of an ONNX repo. The changes in `common.py` include some utilities for reporting and some utilities to create ONNX sessions and input bindings. My 2c is that it's probably fine to include the ONNX session and input binding logic but for the rest probably not. cc @malfet

**Comment by BowenBao** [comment]
Hey @msaroufim thanks for taking a look. I'm ok with moving out the bash scripts and stuff, thought it was more convenient to use by including a simplified interface for launching the onnx benchmark.

We would need some reporting logic checked in though to be able to emit the extra csv logs. From code re-usability and robustness point of view the read/write of the extra csvs are handled by the same class. Would you prefer we blend these into `common.py`, `runner.py`? I'm currently keeping it separated in `_onnx/reporter.py`.

**Review by jansel — APPROVED** [review]
Changes to dynamo benchmarking seem fine to me, not sure which repo is right.

**Comment by malfet** [comment]
> cc @malfet for inputs.

Sorry for the late reply. IMO shell scripts should probably go to `.ci` folder, and also should be written to be more location agnostic (i.e. right now they are written with the assumption that script folder is its CWD, which often is not the case)

Also, I'm not sure how this is currently invoked from CI. If this is part of dynamo shard, than I think it's a bad UX (as dynamo tests are only invoked when certain files are modified), it should be run as part of its own shard (and probably on CPU machine, as it does not need use any GPU resources right now, is it?)
